File: src/main/java/xyz/baddeveloper/lwsl/server/SocketHandler.java
package xyz.baddeveloper.lwsl.server;

import org.json.JSONObject;
import xyz.baddeveloper.lwsl.Packet;
import xyz.baddeveloper.lwsl.server.events.ServerDisconnectEvent;
import xyz.baddeveloper.lwsl.server.events.ServerPacketReceivedEvent;
import xyz.baddeveloper.lwsl.server.events.ServerPacketSentEvent;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.Socket;
import java.util.concurrent.Executors;

public class SocketHandler {

    private SocketServer socketServer;
    private Socket socket;

    private DataInputStream dis;
    private DataOutputStream dos;

    SocketHandler(SocketServer socketServer, Socket socket) {
        this.socketServer = socketServer;
        this.socket = socket;

        try {
            dis = new DataInputStream(socket.getInputStream());
            dos = new DataOutputStream(socket.getOutputStream());
        } catch (IOException ignored) {
        }

    }

    public void handle() {
        Executors.newSingleThreadExecutor().execute(() -> {
            while (!socket.isClosed()) {
                try {
                    Packet packetIn = new Packet(new JSONObject(dis.readUTF()));
                    socketServer.eventManager.fireEvent(new ServerPacketReceivedEvent(socketServer, this, packetIn));
                } catch (Exception e) {
                    try {
                        socket.close();
                    } catch (IOException ignored) {
                    }
                    socketServer.eventManager.fireEvent(new ServerDisconnectEvent(socketServer, this));
                    socketServer.getClients().remove(this);
                    break;
                }
            }
        });
    }

    public void sendPacket(Packet packet) {
        try {
            packet.setName(packet.getClass().getSimpleName().replace("Packet", ""));
            dos.writeUTF(packet.getObject().toString());
            dos.flush();
            socketServer.eventManager.fireEvent(new ServerPacketSentEvent(socketServer, packet));
        } catch (IOException ignored) {
        }
    }

    public Socket getSocket() {
        return socket;
    }

    public DataInputStream getDis() {
        return dis;
    }

    public DataOutputStream getDos() {
        return dos;
    }
}

File: src/main/java/xyz/baddeveloper/lwsl/server/SocketHandler.kt
package xyz.baddeveloper.lwsl.server

import org.json.JSONObject
import reactor.core.publisher.Mono
import reactor.netty.Connection
import xyz.baddeveloper.lwsl.Events
import xyz.baddeveloper.lwsl.packet.Packet
import xyz.baddeveloper.lwsl.server.events.ServerEvent
import xyz.baddeveloper.lwsl.server.events.ServerPacketEvent
import java.io.IOException
import java.net.InetSocketAddress
import java.util.*
import java.util.concurrent.Executors

class SocketHandler internal constructor(private val socketServer: SocketServer, val connection: Connection) {
    var id: String = UUID.randomUUID().toString()

    fun handle() {
        Executors.newSingleThreadExecutor().execute {
            while (connection.channel().isOpen) {
                try {
                    val packetIn = Packet(JSONObject(connection.inbound().receiveObject().subscribe()))
                    socketServer.eventManager.fireEvent(ServerPacketEvent(Events.PACKET_RECEIVED.toString(), socketServer, this, packetIn))
                } catch (e: Exception) {
                    try {
                        connection.dispose()
                    } catch (ignored: IOException) {
                    }
                    socketServer.eventManager.fireEvent(ServerEvent(Events.DISCONNECT.toString(), socketServer, this))
                    socketServer.clients.remove(this.id)
                    break
                }
            }
        }
    }

    fun sendPacket(packet: Packet) {
        try {
            packet.name = packet.javaClass.simpleName.replace("Packet", "")
            connection.outbound().sendString(Mono.just(packet.jsonObject.toString()))
            socketServer.eventManager.fireEvent(ServerPacketEvent("packetSent", socketServer, this, packet))
        } catch (ignored: IOException) {
        }
    }

    fun getRemoteAddress(): InetSocketAddress {
        return connection.channel().remoteAddress() as InetSocketAddress
    }
}